### PR TITLE
Fix building qgis with Qt6 on Windows

### PR DESCRIPTION
**Pull request by MehdiChinoune**
## Description

This fix building qgis with Qt6 on Windows.
Qt6 define `UNICODE` on Windows which makes many functions expect wchar_t input arguments. There are two ways to fix that. The first is to convert all those arguments to wchar_t and the second way is to use ANSI functions explicitly. I have chosen the second way because it's the easiest one.

This also prevents adding `-DWITH_QTWEBKIT` with Qt6 as It breaks compilation.

Also there was a function that wasn't exported for some reason (even It's part of a class, maybe It's a bug or c++17 standard I don't know!) as It gives a linking failure, so I have explicitly exported it.

### CODE REVIEW

**Comment by nicogodet** [comment]
@vcloarec Change on MDAL shouldn't be done on QGIS side, shouldn't it ?

**Comment by vcloarec** [comment]
For minor changes like this, this have to be done, at least, on the both side. So, when MDAL will be updated in QGIS, changes will remain.

**Comment by MehdiChinoune** [comment]
Well, There is a third way which is adding `QT_NO_UNICODE_DEFINES` target property to libraries linked to Qt6 (`set_target_properies(target PROPERTIES QT_NO_UNICODE_DEFINES ON)`) or globally (`set_property(GLOBAL PROPERTY QT_NO_UNICODE_DEFINES ON)`) without modifying any source-file.

File: CMakeLists.txt [hunk 1]
```diff
@@ -401,7 +401,7 @@ if(WITH_CORE)
   endif()
 
   set (WITH_QTWEBKIT TRUE CACHE BOOL "Enable QtWebkit support")
-  if (WITH_QTWEBKIT)
+  if (WITH_QTWEBKIT AND NOT BUILD_WITH_QT6)
```
Comment (m-kuhn) [inline CMakeLists.txt:404]: I think we should instead explicitly fail, can you add a condition on these two and `MESSAGE(FATAL_ERROR "WITH_QTWEBKIT is not supported for Qt6 builds")`.

Comment (MehdiChinoune) [reply CMakeLists.txt:404]: Maybe a WARNING + Forcing it to FALSE, it default to TRUE so It's not user/developer fault!

Comment (m-kuhn) [reply CMakeLists.txt:404]: That or changing the default for Qt6 are both fine for me :+1:

File: CMakeLists.txt [hunk 2]
```diff
@@ -401,6 +401,10 @@ if(WITH_CORE)
   endif()
 
   set (WITH_QTWEBKIT TRUE CACHE BOOL "Enable QtWebkit support")
+  if (BUILD_WITH_QT6)
+    set (WITH_QTWEBKIT FALSE CACHE BOOL "Enable QtWebkit support" FORCE)
+  endif()
```
Comment (m-kuhn) [inline CMakeLists.txt:406]:
```suggestion
  if (BUILD_WITH_QT6)
    set (WITH_QTWEBKIT_DEFAULT FALSE)
  else()
    set (WITH_QTWEBKIT_DEFAULT TRUE)
  endif()
  set (WITH_QTWEBKIT ${WITH_QTWEBKIT_DEFAULT} CACHE BOOL "Enable QtWebkit support")
  
  if (BUILD_WITH_QT6 AND WITH_QTWEBKIT)
    message(FATAL_ERROR "QtWebkit is not supported on Qt6. Disable WITH_QTWEBKIT or BUILD_WITH_QT6.")
  endif()
```

The current proposed approach will also silently fail, which is not transparent to the user. Better change the default.

**Comment by m-kuhn** [comment]
I'm fine with the webkit and export change.

On the unicode front, as far as I understand this will use 8bit strings interpreted in system locale which seems to be calling for encoding issues. And the `wchar_t` approach would be safer. I'm not an expert so happy to be proven wrong.

**Comment by MehdiChinoune** [comment]
> I'm fine with the webkit and export change.
> 
> On the unicode front, as far as I understand this will use 8bit strings interpreted in system locale which seems to be calling for encoding issues. And the `wchar_t` approach would be safer. I'm not an expert so happy to be proven wrong.

I agree, but it requires more changes.

**Comment by MehdiChinoune** [comment]
QGIS is already Non-UNICODE, what's the issue of keeping.
Is it necessary to fellow Qt6.
This is not something that will break existing encoding, It just keeps it.

**Comment by m-kuhn** [comment]
I've seen encoding errors in message boxes and other places on windows. I think it's good to do this the clean way and improve the integration of this platform. Otherwise someone else will have to do it and find all affected places sooner or later. I think it's a unique chance we are pointed to all these problematic places by the compiler.

**Pull request closed without merging**